### DISCUSSION TITLE
How to change icon

### DISCUSSION THREAD

**Discussion by AlexsandroRP** [TomSchimansky/CustomTkinter#1099] [General]
Hey Tom, hello. How can i change the default window icon? I tried a lot of regular tkinter methods 
unsuccessfully. Thanks!

**Comment by IAminos** [comment]
Hello,

I had the same issue, I had this "fix" for it: 

`root.after(201, lambda :root.iconbitmap('C:\\Users\\...\\filename.ico'))` instead of  `root.iconbitmap('C:\\Users\\...\\filename.ico')`

I hope someone will provide a better solution

**Reply by RobbieKhan** [reply]
I can confirm that currently that is the only one solution that works

**Reply by saiyam-sandhir** [reply]
could you please answer how you came up with the number 201 (time) cause I found that it wasn't working for me with any other number

**Reply by soukisami** [reply]
thx bro it worked

**Reply by amezolaMartin** [reply]
It worked! Thaks a lot.

**Reply by Maaz4226** [reply]
What do we do when we inherit a class with self?

**Reply by Maaz4226** [reply]
![Screenshot 2024-04-13 125123](https://github.com/TomSchimansky/CustomTkinter/assets/160339696/0c24c057-f403-4b08-8e99-e04350f7b93f)

**Reply by Akascape** [reply]
@Maaz4226 
replace the root with self.window in your case

**Reply by Maaz4226** [reply]
> @Maaz4226 replace the root with self.window in your case

Your code is correct but I wrote a lot of code with self so I wrote this code "self.window.iconbitmap()"

**Comment by TomSchimansky** [comment]
If you use the latest version there should be no problem.

**Comment by AlexsandroRP** [comment]
@IAminos Thanks a lot, that was the only method that worked!

**Comment by Tcrozly** [comment]
This issue still effects CTkToplevel, using the suggestion from IAminos solved my issue. thanks!

**Comment by liquidz00** [comment]
Has anyone gotten this to cooperate on macOS? I made sure the latest version of CustomTkinter is running. Using the regular `app.iconbitmap(PATH_HERE)` does not cooperate, nor does the solution presented by IAminos :(

**Reply by Akascape** [reply]
@aspeciale I recommend using the `iconphoto` method. It works on every platform.

**Reply by avalon60** [reply]
@Akascape I use this, but it only seems to work on Linux:
```

        icon_photo = tk.PhotoImage(file=APP_IMAGES / 'bear-logo-colour-dark.png')
        self.iconphoto(False, icon_photo)
```
![image](https://github.com/TomSchimansky/CustomTkinter/assets/89534395/64091010-987e-4fb3-9183-91101142cc8d)

Windows:

![image](https://github.com/TomSchimansky/CustomTkinter/assets/89534395/36497aa3-a7af-4459-a348-a7dd4f976c36)

Any ideas what I am doing wrong?

**Reply by avalon60** [reply]
Actually, I stand corrected. I only get the icons when running from PyCharm. If I run standalone, I get:

![image](https://github.com/TomSchimansky/CustomTkinter/assets/89534395/bf556f1a-dcca-4917-91e1-76bc6914bc61)

**Reply by liquidz00** [reply]
I'm a bit of a novice with coding, you'll have to forgive me. I assume you mean `app.iconphoto(<image variable here>)` assuming `app = customtkinter.CTk()`? Would you be able to provide an example of the method implementation?

**Reply by avalon60** [reply]
Well, in my example, you need to assume that I am in the context of a class based on CTk(), or CTkTopLevel. So instead of app... I am referencing it as self...

That said, it appears you might be missing a parameter. See [here](https://www.geeksforgeeks.org/iconphoto-method-in-tkinter-python/).

**Reply by Akascape** [reply]
@avalon60 We cannot change that default python taskbar icon on windows. But if you compile it to exe, then it will work properly.

**Reply by liquidz00** [reply]
@avalon60 Going to be honest I hyper-focused on @Akascape 's response and glossed over your reply, I didn't realize you were also using said method. 

I'm using classes, both application-specific and inheriting from customtkinter. I originally tried setting that property on self, but the issue persisted.  I have since put this to the side as I'm curious to see in testing what PyInstaller can do instead. Thank you though!

**Reply by Akascape** [reply]
@aspeciale Please try this method: https://github.com/TomSchimansky/CustomTkinter/discussions/1099#discussioncomment-6507015

**Reply by gulaki** [reply]
On MacOS I use py2app to set the icon file in my `setup.py`

```
setup(
  'app': 'path/to/myscript.py`,
  'options': {
    'py2app': {
      'iconfile': 'my-icon',
  }})
  ```

  Running `python setup.py py2app` builds a Mac app bundle. This sets the app icon in the bundle as well as during runtime in the taskbar, window title, etc. I feel this is a stable option.

**Note:** Don't use conda to build the bundle, use the official distribution from www.python.org.

**Comment by Akascape** [comment]
**Try this method:**

`For root window:`
```python
import os
from PIL import ImageTk
...

self.iconpath = ImageTk.PhotoImage(file=os.path.join("assets","logo.png"))
self.wm_iconbitmap()
self.iconphoto(False, self.iconpath)
```
`For toplevels:`
```python
self.wm_iconbitmap()
self.after(300, lambda: self.iconphoto(False, self.iconpath))
```

**Reply by liquidz00** [reply]
The root window did the trick! Thank you so much @Akascape !

**Reply by SaimNaseem** [reply]
@TomSchimansky Hi Tom , wie kann ich meine GUI 'logo' wechseln mit customtkinter. Es ist unverbesserlich wann ich die Gleiche Logo für Jede Python GUI habe.

**Comment by bukdk** [comment]
import customtkinter as ctk
import tkinter as tk
import sys
import os
from pathlib import Path
from PIL import ImageTk 
from pathlib import Path
 
class App(ctk.CTkFrame):
    def __init__(self, master):
        super().__init__(master)

        self.master = master

        # Disable the Close Window Control Icon depends on the operating system
        if sys.platform == "linux":
            self.overrideredirect(True)
        elif sys.platform == "darwin":  # OS X
            pass
        else:  # win32
            self.master.protocol("WM_DELETE_WINDOW", self.disable_event)
     
        self.pfad = os.path.dirname(__file__)
        os.chdir(self.pfad)
        self.iconpath = ImageTk.PhotoImage(
            file=os.path.join("resources/images", "logo.ico")
        )
        self.master.wm_iconbitmap()
        self.master.iconphoto(False, self.iconpath)
        # but for a CTkToplevel you have to change self.master.iconphoto(False, self.iconpath) to
        # self.after(300, lambda: self.iconphoto(False, self.iconpath))

        self.create_widgets()

    def create_widgets(self):
        '...'


root = ctk.CTk()
root.title("Test App")
root.fg_color = "#161C25"
app = App(root)
root.eval(f"tk::PlaceWindow {str(root)} center")
app.mainloop()



This worked for me (KDK)

**Comment by ghost** [comment]
Something like this:
```py
# Importing required module
import customtkinter as ctk

# Selecting GUI theme : dark,light , system (for system default)
ctk.set_appearance_mode("system")

# Selecting color theme-blue, green, dark-blue
ctk.set_default_color_theme("dark-blue")

app = ctk.CTk()
app.geometry("400x400")
app.title("Modern Login UI using Customtkinter")
openstack_icon = app.iconbitmap('openstack.ico')
```

**Reply by SphinX2k6** [reply]
yes but how do i compile it with pyinstaller

**Reply by Haolinc** [reply]
> yes but how do i compile it with pyinstaller

do something like this to get pyinstaller path:
```
if getattr(sys, 'frozen', False) and hasattr(sys, '_MEIPASS'):
        return os.path.join(sys._MEIPASS, file_name)
    else:
        return os.path.join(os.path.abspath(os.getcwd()), file_name)
```
then when calling pyinstaller add this piece of command along with rest of your pyinstaller one liner command:
`--add-data="your_icon.ico:your_icon.ico"`

**Comment by wilsondesouza** [comment]
---

For who have tried the solutions above, but still have the problem (the solutions only work when you run the code, not when create an exe).

Use this in your code
`...wm_iconbitmap("absolute/path/yourico.ico")`

Something like that
```
self.wm_iconbitmap("absolute/path/yourico.ico") # <-- What you want (absolute path)
self.title("Your Application Name")
self.geometry("800x700")     
self.minsize(650, 600)
```

Save the code and get your `customtkinter` absolute path lib with
`pip show customtkinter` 

And then, finally put your `pyinstaller` command:
`pyinstaller --noconfirm --onefile --name "Your Application Name" --windowed --add-data "path/to/Lib/site-packages/customtkinter;customtkinter/" --icon="path/to/yourico.ico" "yourcode.py"`

&#8593; Should work with relative and absolute path

---

**Comment by jnbutler1815** [comment]
None of these methods worked for me on Zorin OS 18 Pro.
